### ISSUE TITLE
clang format

### ISSUE THREAD

**Issue by tony** [mawww/kakoune#723]
Any thoughts on [clang format](http://clang.llvm.org/docs/ClangFormat.html)?  It is used at google on pretty big projects. The output is pretty consistent and its very configurable.

My argument against clang-format

- someone can eye-ball just about any code and make it somewhat consistent.
- it doesn't always preserve manual spacing / padding very well, such as  https://github.com/martanne/vis/blob/8696abb/config.def.h#L27. I haven't found a way to keep this other than wrapping in ``// clang-format off`` and ``// clang-format on``
- Initial wariness of formatters in general over the past 10 years. In many cases I've always ended up doing it by hand and formatting tended to ironically lead to less consistency than more.

My argument in favor of:

- Out of any formatter I ever tried, this one can run one command and give me consistent code everywhere. Even in cases where it wasn't a familiar style, including outright jibberish/random config settings, I found it the formatting alone made the project more readable since the results is reliably homogenous project-wide
- ``.clang-format`` can set formatting configuration on a per-project basis. 
There is a config utility at http://zed0.co.uk/clang-format-configurator where you can get a live preview of how config settings affect code
- LLVM, Chromium, Google projects depend on it, as well as a few friends I know use it at work and give it pretty glowing reviews.
- configuration can be pretty granular, I can have it adhere to formatting things in https://github.com/tony/kakoune/blob/cmake/doc/coding-style.asciidoc#c-coding-style

As an example, I provided a minimal configuration that should satisfy our coding style. If you used this ``.clang-format`` on clang-format 3.8:

```clang-format
---
BreakBeforeBraces: Stroustrup
ColumnLimit: '80'
IndentWidth: '4'
Language: Cpp
UseTab: Never
```

<img width="1495" alt="screen shot 2016-07-03 at 11 53 35 am" src="https://cloud.githubusercontent.com/assets/26336/16546588/cfd1f35a-4114-11e6-8002-cb0b467a3da4.png">

**Comment by mawww** [comment]
Hello,

My general opinion on the matter is that code formatters are not good enough yet to handle various corner cases, in particular when in comes to lambdas, column limits, arguments... I actually doubt they could ever become good enough, because there are some informations I tend to use (from time to time) they dont have access to. The main example for that would be splitting arguments in a function definition where they go above the limit:

```
DisplayAtom(const Buffer& buffer, ByteCoord begin, ByteCoord end);
```

Here, if I needed to split the line, I would probably the argument list after the buffer argument, to keep the begin and end argument grouped together, instead of splitting after the begin argument as a formatter would probably do.

For this reason, I tend to prefer to leave the last word on formatting to whoever writes the code, instead of requiring the code to go through a formatter.

In other words, formatters do a good enough job most of the time, but I like to keep control for the cases where they dont do a good job, so I wont be adding a git hook to ensure we go through clang-format or anything like that. And i find adding comments to disable/reenable the code formatter to be too ugly to consider it.

Cheers,

Maxime.

**Comment by tony** [comment]
>  I actually doubt they could ever become good enough, because there are some informations I tend to use (from time to time) they don't have access to.

It is just styling in the end.  If its hard for a formatter to do, how would a personal styling touch or a specialized styling directive last in the long run over many different contributors/readers, changes and a bigger codebase? If its a special-case to the styling everywhere else in the codebase, it could hinder a new person's ability to read. That's normally the market I'm catering to.

I haven't found a way in clang-format (up to HEAD) to get what you would like in your example.

I assume here: ``DisplayAtom(const Buffer& buffer, ByteCoord begin, ByteCoord end);``

The desired effect would be:

```
DisplayAtom(const Buffer& buffer, 
                     ByteCoord begin, ByteCoord end);
```

The first option people would pick would be ``BinPackArguments: False``:

```
  DisplayAtom(const Buffer &buffer,
              ByteCoord begin,
              ByteCoord end);
```

If I understand correctly, what you would want is:

```
  DisplayAtom(const Buffer &buffer,
                       ByteCoord begin, ByteCoord end);
```

That is actually a good idea for a clang-format option. There could be a patch made to preserve cases of a first argument on the first line, and behave like BinPack in lines thereafter. And languages like C would like it, since the first argument is frequently used to pass around a context.

> And i find adding comments to disable/reenable the code formatter to be too ugly to consider it.

That would be for edge cases, namely that specific example. In other cases, its a conscious decision to forgo the human touches.

Will that rule out the other patch I had to wrap server functionality in ``#ifdef`` so it can optionally not be compiled?

>  so I wont be adding a git hook to ensure we go through clang-format or anything like that.

I would be wary of that. I don't imagine those who are successful with clang-format will use it in that fashion.

I doubt clang-format users just chuck it in a cargo-cult way. I've been accused tangentially of doing this with linting in interpreted languages (binding linters to a git hook, making lint count go to 0). But in the end, it helps me catch things I otherwise overlook.  Honestly, I just juggle a lot of projects, I'm doing what I can to keep the quality high while manage other things.

I think that its a "net effect" question. I imagine when chromium opts to go for clang-format, they're abandoning a lot of human touches in favor of uniformity. But one person's permission of ugliness/style is mostly their own.

> For this reason, I tend to prefer to leave the last word on formatting to whoever writes the code, instead of requiring the code to go through a formatter... but I like to keep control for the cases where they dont do a good job

I don't know what "good job" is in this sense. But if it were found clang-format were to give jibberish, or even potentially something that wouldn't compile, I'd say no to clang-format.  If good styling was guaranteed 98% of the time, consistently, even at the cost of losing 2% of special touches that were merely my own personal habit, I'd go clang-format every time. If that 2% was universally unreadable/inconsistent with the rest of the 98% in really critical areas, I wouldn't use it.

Big projects often go 10+ years hardening their own stylistic taste.  That means they evolve a very engrained sense of what looks good and readable, but then will find another respected C++ project with totally different tastes, to see http://zed0.co.uk/clang-format-configurator/, the top configuration has a ``BasedOnStyle`` option of Default, LLVM, Chromium, Google, Webkit.

If you don't want to proceed with clang-format, you can close this.

**Comment by doppioandante** [comment]
Kakoune currently enjoys the status of "small project", with mawww being the main code contributor.
For the moment I don't think that adding an external tool for formatting would be so beneficial, the pull request review process should be enough to spot blatantly ugly code.
I also don't mind the concept of "ugly code" to be subjective. Personally I find the codebase quite clean, and mawww has never been picky on style as far as I remember.

As kak follows the UNIX philosophy, I would say that the code base should not go beyond a certain size.
This is questionable though. In case kakoune ever becomes a big project with many contributors, then I agree that a formatting tool would be very useful. Let's not higher the contributing treshold for now.

**Comment by tony** [comment]
> As kak follows the UNIX philosophy, I would say that the code base should not go beyond a certain size.

I'm fine with both of those, but how does adding a ``.clang-format`` file affect that? Its more of a meta-issue.

TBH it affects me directly because I have a fork, prefer to use clang-format (regardless of the settings kakoune would use *if* it used a ``.clang-format`` config), and merging upstream changes / PR'ing patches back to kakoune gets much more difficult. :P I have to choose between not using clang-format  or having to go through a lot more labor every time I want to submit a patch/sync things.

> Let's not higher the contributing treshold for now.

How / what makes contributing more difficult? I would anticipate the opposite, the biggest grievance would be for contributors, its really more a favor to new contributors to have consistency like that.  Even if I purposely make the formatting look unusual as possible, its at least homogeneously unusual. :smile:

**Comment by lenormf** [comment]
Add `.clang-format` to the `.gitignore` file, never add `.gitignore` to any commit, and that should be it. You could also keep a branch with your local modifications in a few commits, rebase your feature branch on it, and then rebase the branch back before pushing it for review.

No need to push specific tools on everyone, otherwise it's going to open a can of worm that will be impossible to close back.

**Comment by tony** [comment]
> Add .clang-format to the .gitignore file, never add .gitignore to any commit, and that should be it.

clang-format formats the code.  the ``.clang-format`` file would be the project's.

if clang-format isn't being used, then there won't be any formatting with it, and no need for a ``.clang-format`` file.

You could also keep a branch with your local modifications in a few commits, rebase your feature branch on it, and then rebase the branch back before pushing it for review.

I have 5 branches or so branches. In one of my earliest ones, I ran clang-format, and it made it extremely painful to rebase against the latest changes in either direction.

> No need to push specific tools on everyone,

Oh fantastic, now I'm the big tyrannical cargo culting monster! :smile:  You try forking this project, trying to slim it down for another use, then being left to deal with singletons and multiple base classes "pushed on" you. Then have somebody tell you that. 😞 

However, it may not be a problem if kakoune isn't the project I fork.

This is purely a meta issue intended to make everyone's lives easier. Its C++ and my own personal taste, acquired from larger projects that have shown success with it.

That said, I don't want to drag this out.  I highly recommend giving clang-format a shot some time if you have free time.

**Comment by lenormf** [comment]
>This is purely a meta issue intended to make everyone's lives easier. Its C++ and my own personal taste, acquired from larger projects that have shown success with it.
>That said, I don't want to drag this out. I highly recommend giving clang-format a shot some time if you have free time.

You completely missed the point, everyone has their own preferences as to what tools they use, there's no need to prioritize one over all the others, especially since it's so easy for you to use it on your side without adding extra files in the repository. The only exceptions to that is the `contrib` directory, which could probably hold just about any configuration file you want as long as the tool is objectively useful (a homebrew recipe, a different build system's build file etc).

You've also been given an example that shows that the tool might not fit all of a developer's needs, which cuts in half any further consideration to include it by default.

**Comment by tony** [comment]
> everyone has their own preferences as to what tools they use, 

Its a tool, but when you adopt it, your letting it project's clang-format configuration and the tool the shots for styling.  It doesn't necessarily require outside contributor to download the tool.

> especially since it's so easy for you to use it on your side without adding extra files in the repository. 

No, I would simply not use it all, because:

https://github.com/Microsoft/CNTK/commit/2f39fe3f01aaabae19a6e8dec3ae859fb61da10b

It'll change every file. And it depends on the settings adopted, as they are consistently applied throughout.

A project adopts clang-format's styling or it doesn't. Post-processing is possible.  You could technically use it over an area of code, if there isn't a commonly adopted clang-format configuration, it will look drastically different than the surround code.
